include/impl_ptr.hpp

```diff
@@ -112,7 +112,10 @@ struct impl_ptr
 
     template<typename, template<typename, typename...> class, typename...> friend struct impl_ptr;
 
-    impl_ptr(std::nullptr_t) : impl_(nullptr) {}
+    impl_ptr(std::nullptr_t) : impl_(nullptr, typename policy_type::allocator_type()) {}
+
+    template <typename allocator, typename = typename std::enable_if<std::uses_allocator<policy_type, allocator>::value>::type>
+    impl_ptr(std::nullptr_t, allocator&& a) : impl_(nullptr, std::forward<allocator>(a)) {}
```

Comment:
I understand the reasoning why you added allocator to the impl_ptr(std::nullptr_t) constructor... but don't you find that weird? After all, impl_ptr(null) does not allocate anything... Say, shared_ptr comes to mind.  So, IMO passing an allocator in this instance feels counter-intuitive and confusing. Don't you agree?

Reply:
Sorry for the delayed reply.

In short: in general, during the lifetime of an object (any object) you want it to have the same allocator instance. This enforces that, to a degree. It took me a long time to try and figure out how to do things correctly _without_ that, but I couldn't get to anything that felt highly error prone.

I'd suggest the C++Now talk about std::pmr (polymorphic memory resource) by David Sankel. It explains, more clearly than I can, why you want this kind of behaviour (more accurately, why you _don't_ want changes in allocator type).

https://www.youtube.com/watch?v=FLbXjNrAjbc